File: python/problems/leetcode/0380_insert_delete_getrandom_o1/solution.py
"""LeetCode 380: Insert Delete GetRandom O(1)."""

from __future__ import annotations

import random


class RandomizedSet:
    def __init__(self) -> None:
        self._idx: dict[int, int] = {}
        self._vals: list[int] = []

    def insert(self, val: int) -> bool:
        if val in self._idx:
            return False
        self._idx[val] = len(self._vals)
        self._vals.append(val)
        return True

    def remove(self, val: int) -> bool:
        if val not in self._idx:
            return False
        i = self._idx[val]
        last = self._vals[-1]
        self._vals[i] = last
        self._idx[last] = i
        self._vals.pop()
        del self._idx[val]
        return True

    def getRandom(self) -> int:
        return random.choice(self._vals)


if __name__ == "__main__":
    s = RandomizedSet()
    assert s.insert(1)
    assert not s.insert(1)
    assert s.remove(2) is False
    assert s.insert(2)
    assert s.getRandom() in {1, 2}
    assert s.remove(1)
    assert s.getRandom() == 2
    print("0380 OK")

File: cpp/problems/leetcode/0380_insert_delete_getrandom_o1/solution.cpp
// LeetCode 380: Insert Delete GetRandom O(1)
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class RandomizedSet {
    unordered_map<int, int> idx;
    vector<int> vals;

public:
    bool insert(int val) {
        if (idx.count(val)) return false;
        idx[val] = (int)vals.size();
        vals.push_back(val);
        return true;
    }

    bool remove(int val) {
        if (!idx.count(val)) return false;
        int i = idx[val];
        int last = vals.back();
        vals[i] = last;
        idx[last] = i;
        vals.pop_back();
        idx.erase(val);
        return true;
    }

    int getRandom() {
        return vals[rand() % vals.size()];
    }
};

int main() {
    RandomizedSet s;
    assert(s.insert(1));
    assert(!s.insert(1));
    assert(!s.remove(2));
    assert(s.insert(2));
    s.remove(1);
    assert(s.getRandom() == 2);
    cout << "0380 OK" << endl;
    return 0;
}
